src/osd/PrimaryLogPG.cc

```diff
@@ -8313,6 +8324,23 @@ int PrimaryLogPG::_rollback_to(OpContext *ctx, OSDOp& op)
     block_write_on_degraded_snap(missing_oid, ctx->op);
     return ret;
   }
+  /*
+   * In rollback, if the head object is not manfest and the rollback_to is manifest,
+   * the head object becomes the manifest object. At this point,
+   * we need to check adjacent clonces to the head object to calculate 
```

Comment:
clones

Reply:
Why don't we have to recalculate this if the head isn't a manifest object?

Reply:
I don't think I understand your comment. Do you mean that the new head needs to be recalculated?
We have to calculate the new head's reference here (because it is now the manifest). To do so, we need to check whether adjacent clones are readable and wait for the recovery if the adjacent clones are unreadable, then calculate the correct reference.

Reply:
Even if the head object is already a manifest object, it might have no references in common with the rollback target.  It seems to me like this means that we always have to check the adjacent clone.

Reply:
If the head object is already a manifest object, recover_adjacent_clones() in do_op() checks the adjacent clone and calls wait_for_unreadable_object() if necessary. The reason why I added these lines here is that the recover_adjacent_clones() only checks the adjacent clones if the object is manifest. 
I'm not sure that the recover_adjacent_clones() in do_op() is a right place to check the adjacent clone if the head object is not manifest because we need to get the rollback target's object context to see the target snapshot is manifest.

Do you think this should be moved to the recover_adjacent_clones() in do_op() or vice versa?

Reply:
The two cases seem symmetric, so I expect them to be handled in the same manner.  What wait-list gets used in the other case?

Reply:
Right. It would be better to handle the two cases in _rollback_to(). This means that `recover_adjacent_clones() `in `do_op() `does not handle rollback op. Then, checking adjacent clones is done at `_rollback_to()`. 
This way can avoid decoding rollback op's snapid at `do_op()`, moreover it is how other wait-lists (e.g., objects_blocked_on_degraded_snap) handle waiting events.

Reply:
Sounds like a plan.

Reply:
Updated.